[DPE-6430] Add basic status handling

reneradoi opened the pull request:
With this PR, we add basic status handling for units and application. The implementation adheres to the guidelines provided in [Evaluating charm status across multiple components](https://discourse.charmhub.io/t/how-to-set-a-charms-status/11771#heading--evaluate-charm-status-across-multiple-components).

The changed behaviour can be nicely been seen in the integration test runs on CI, for example [this one](https://github.com/canonical/charmed-etcd-operator/actions/runs/13721304582/job/38377439306?pr=50#step:29:786) from line 785 onwards.

The main changes are:

**charm.py:**
- keeping a list of `pending_inactive_statuses`
- observing the `collect_unit_status` and `collect_app_status` events
- do not set status (`TLS_TRANSITION_FAILED`) if we raise an error in the event handler

**EtcdEvents:**
- do not set `ActiveStatus` as this is now part of `CollectStatusEvent` handler
- set status if deferring `on_start` because TLS is not ready or the unit did not join the cluster yet
- set status when encountering errors in `update_admin_password()`

**ClusterManager:**
- new method `compute_cluster_status()`

unrelated changes:
- `literals.py` re-arrange `Status` class in alphabetical order for better readability
- update `tls_certificates` lib from version `4.9` to `4.11`

reneradoi committed:
add observers for `CollectStatus` events, add list of pending statuses to the charm

reneradoi committed:
set status if unit did not yet join the cluster on startup

reneradoi committed:
adjust unit test

reneradoi committed:
adjust unit test

reneradoi committed:
we can not set status if we raise

reneradoi committed:
add feedback to users in case password update fails

reneradoi committed:
set status if tls not ready or cluster member not yet promoted

reneradoi committed:
set status if tls not ready or cluster member not yet promoted

reneradoi committed:
fix linting

reneradoi committed:
update tls lib

reneradoi committed:
add compute logic for collect status event

reneradoi committed:
move logic for computing status to `compute_component_status` in `ClusterManager`

Mehdi-Bendriss commented on src/managers/cluster.py:
```diff
@@ -316,3 +317,16 @@ def update_cluster_member_state(self) -> None:
+        status_list = []
+
+        if self.state.unit_server.is_started:
+            if not self.state.cluster.cluster_state == EtcdClusterState.EXISTING.value:
```
```suggestion
            if self.state.cluster.cluster_state != EtcdClusterState.EXISTING.value:
```

Mehdi-Bendriss commented on src/charm.py:
```diff
@@ -209,6 +208,27 @@ def _restart_clean_cas(self, _) -> None:
+        """
+        # compute cluster status
+        for status in self.cluster_manager.compute_component_status():
+            event.add_status(status.value.status)
```
in case this component takes priority over the following ones, we should perhaps return on the first existing status?

reneradoi replied:
Returning on the first existing status would mean that other, perhaps more important statuses would not be collected. I would try to avoid that.

According to the `CollectStatusEvent` [docs](https://ops.readthedocs.io/en/latest/reference/ops.html#ops.CollectStatusEvent), "If there are multiple statuses with the same priority, the first one added wins". Meaning:
- the order of status priorities (blocked, maintenance, waiting, active) is defined by ops
- the order of component priorities is defined by the order here in this method
- by collecting all component statuses, we make sure to not miss any whose status priority might be ranked higher

In my point of view, this is a very simple, yet comprehensive logic.

Mehdi-Bendriss replied:
Okay, that sounds good. We can explore more advanced status prioritization once the status impl. spec lands

Mehdi-Bendriss reviewed (commented):
Thanks René! I believe it's going on the right direction of the overall status vision we have

reneradoi committed:
Merge branch '3.5/edge' into add-basic-status-handling

reneradoi committed:
apply feedback

skourta committed:
[DPE-6790] - TLS Status Handling (#53)

This pull request includes several changes focused on improving the
handling and computation of TLS statuses within the `etcd` charm. The
changes involve updating the methods for setting and computing TLS
statuses, as well as modifying the status levels and their logging
levels.

Improvements to TLS status handling:

*
[`src/charm.py`](diffhunk://#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR224-R225):
Added logic to compute and add TLS component statuses in the
`_on_collect_status` method.
*
[`src/events/etcd.py`](diffhunk://#diff-4d4b45129d72bf79a08e70009a59f4ebe0a016b4727117155249c864ae4a1b49L106):
Removed the direct setting of TLS status and replaced it with deferring
events and adding component statuses through the `tls_manager`.
[[1]](diffhunk://#diff-4d4b45129d72bf79a08e70009a59f4ebe0a016b4727117155249c864ae4a1b49L106)
[[2]](diffhunk://#diff-4d4b45129d72bf79a08e70009a59f4ebe0a016b4727117155249c864ae4a1b49L363-R362)
*
[`src/events/tls.py`](diffhunk://#diff-3dfb4a104b407f7d4b2e07795a63745a09cde228a8dbe86602e7b72ea600fbe4L99-R105):
Updated the initialization and event handling methods to use the
`tls_manager` for adding component statuses instead of directly setting
the status.
[[1]](diffhunk://#diff-3dfb4a104b407f7d4b2e07795a63745a09cde228a8dbe86602e7b72ea600fbe4L99-R105)
[[2]](diffhunk://#diff-3dfb4a104b407f7d4b2e07795a63745a09cde228a8dbe86602e7b72ea600fbe4L159-L162)

Status level modifications:

*
[`src/literals.py`](diffhunk://#diff-959ba637043ab4ac88c6f9f9109022f78a7df1b9a2ab3a0407bff4f124c2db64L86-R90):
Changed the status level for `TLS_NOT_READY` to `MaintenanceStatus` with
a "DEBUG" log level and added new statuses for peer and client CA
rotations.

Enhancements to `tls_manager`:

*
[`src/managers/tls.py`](diffhunk://#diff-a2c036f44ae9bbab595aad84b23bc3106a342c5ab12f7b0cf3d26268b51cdc83L17-R17):
Introduced a list to track component statuses, and added methods to
compute component statuses and add new statuses to the list.
[[1]](diffhunk://#diff-a2c036f44ae9bbab595aad84b23bc3106a342c5ab12f7b0cf3d26268b51cdc83L17-R17)
[[2]](diffhunk://#diff-a2c036f44ae9bbab595aad84b23bc3106a342c5ab12f7b0cf3d26268b51cdc83R29)
[[3]](diffhunk://#diff-a2c036f44ae9bbab595aad84b23bc3106a342c5ab12f7b0cf3d26268b51cdc83R205-R236)

---------

Co-authored-by: reneradoi <rene.radoi@canonical.com>
